api/py/ai/chronon/repo/validator.py

```diff
@@ -175,18 +241,40 @@ def _validate_join(self, join: Join) -> List[str]:
                          if self._get_old_obj(GroupBy, group_by.metaData.name)]
         non_prod_old_group_bys = [group_by.metaData.name for group_by in old_group_bys
                                   if group_by.metaData.production is False]
+        # Check if the underlying groupBy is valid
+        group_by_errors = [self._validate_group_by(group_by) for group_by in included_group_bys]
+        errors += [f"join {join.metaData.name}'s underlying {error}"
+                   for errors in group_by_errors for error in errors]
+        # Check if the production join is using non production groupBy
         if join.metaData.production and non_prod_old_group_bys:
             errors.append("join {} is production but includes the following non production group_bys: {}".format(
                 join.metaData.name, ', '.join(non_prod_old_group_bys)))
+        # Check if the online join is using the offline groupBy
         if join.metaData.online:
             if offline_included_group_bys:
                 errors.append("join {} is online but includes the following offline group_bys: {}".format(
                     join.metaData.name, ', '.join(offline_included_group_bys)))
-            # If join is online we materialize the underlying group bys
-            # So we need to check if they are valid.
-            group_by_errors = [self._validate_group_by(group_by) for group_by in included_group_bys]
-            errors += [f"join {join.metaData.name}'s underlying {error}"
-                       for errors in group_by_errors for error in errors]
+        # Only validate the join derivation when the underlying groupBy is valid
+        group_by_correct = all(not errors for errors in group_by_errors)
+        if join.derivations and group_by_correct:
+            columns = set(get_pre_derived_join_columns(join))
+            derived_columns = set()
+            for derivation in join.derivations:
+                if derivation.name in derived_columns:
+                    errors.append(
+                        "Incorrect derivation name {} due to output column name conflict".format(derivation.name))
+                else:
+                    derived_columns.add(derivation.name)
+            for derivation in join.derivations:
+                dev_name = derivation.name
+                dev_exp = derivation.expression
+                if dev_name in columns:
+                    errors.append("Incorrect derivation name {} due to output column name conflict".format(dev_name))
+                if dev_exp != "*" and is_identifier(dev_exp):
+                    if dev_exp not in columns:
+                        errors.append(
+                            "Incorrect derivation expression {}, please check the derivation expression".format(
+                                dev_exp))
```

Comment:
we could refactor this code block with to take derivations as an input param method.